internal/events/sql.go

```diff
@@ -27,11 +28,24 @@ const (
 	);`
 	dropEventsTable = `DROP TABLE IF EXISTS ` + schemaName + `.events;`
 
-	insertEventIdempotent  = `INSERT INTO ` + schemaName + `.events (id, data, event_type) VALUES ($1, $2, $3) ON CONFLICT DO NOTHING;`
-	deleteEvent            = `DELETE FROM ` + schemaName + `.events WHERE id = $1;`
-	getEvents              = `SELECT data, event_type FROM ` + schemaName + `.events;`
-	getUnbroadcastedEvents = `SELECT data, event_type FROM ` + schemaName + `.events WHERE NOT received AND NOT broadcasted;`
-	markReceived           = `UPDATE ` + schemaName + `.events SET received = TRUE WHERE id = $1;`
+	insertEventIdempotent = `INSERT INTO ` + schemaName + `.events (id, data, event_type) VALUES ($1, $2, $3) ON CONFLICT DO NOTHING;`
+	deleteEvent           = `DELETE FROM ` + schemaName + `.events WHERE id = $1;`
+
+	deleteEvents = `DELETE FROM ` + schemaName + `.events WHERE id =ANY($1);`
+	getEvents    = `SELECT data, event_type FROM ` + schemaName + `.events;`
+
+	getNewEvents = `SELECT e.data, e.event_type
+	FROM ` + schemaName + `.events AS e
+	LEFT JOIN ` + voting.VotingSchemaName + `.resolutions AS r ON e.id = r.id
+	WHERE r.id IS NULL;`
+
+	getUnbroadcastedEvents = `SELECT data, event_type FROM ` + schemaName + `.events WHERE NOT broadcasted;`
+
+	// FilterObservedEvents returns the list of events that are observed by the node and not yet broadcasted.
+	filterObservedEvents = `SELECT unnested.id
+	FROM unnest($1::BYTEA[]) AS unnested(id)
+	LEFT JOIN ` + schemaName + `.events AS res ON unnested.id = res.id
+	WHERE res.id IS NOT NULL AND res.broadcasted = FALSE;`
```

Comment:
I think we should seriously consider performing the set difference in the application.  I don't see a better way to actually get the unobserved events from an input array with just one SQL query, but it's clearly a much simpler query to just get a list of the event ids that are observed and in the application do the filter to be left with the ones that aren't in the table.

Reply:
I would maybe disagree, since the event store could grow fairly large. But why don't you just do:

```sql
SELECT id FROM schema.events WHERE id =ANY($1) AND res.broadcasted = FALSE;
```

Reply:
This is my confusion - man the input `$1` contain IDs that are NOT not the table?

Reply:
But to clarify my comment "consider performing the set difference in the application" I don't mean grab all the events back to the app.  Still filter by id (with `=ANY` if it really should NOT be an outer join), but if it really is possible to have events _not in the table_ and we want those included (as outer join suggest) you can deal with that in the app.